### ISSUE TITLE
gclient sync fails when running nw_patch hook

### ISSUE THREAD

**Issue by KanomiDim** [nwjs/nw.js#7530]
Thanks for reporting an issue! Spend 60 seconds reading this...

1. Before reporting, try reproducing against the latest version of NWJS.
2. Issue tracker is only for bugs and feature requests, in English only.
  
  Anything else? Use our mailing list for questions: 
  https://groups.google.com/forum/#!forum/nwjs-general

======== TEMPLATE BUG FORM ========

NWJS Version : 47.1 (Latest)
Operating System : Windows 10 x64

### Expected behavior

The gclient sync hooks will succeed

### Actual behavior
I get this message:

Person@DESKTOP-F7DHCJS MINGW64 /c/nwjs471
$ gclient sync --with_branch_heads
Syncing projects:   9% ( 9/94) src/third_party/blink/web_tests/wpt_internal/webgpu/third_party/glslang_js:chroSyncing projects:  10% (10/94) src/buildtools/third_party/libc++/trunk                                        Syncing projects: 100% (102/102), done.
Running hooks:  40% (38/94) nw_patch
________ running 'vpython.bat src/content/nw/tools/patcher.py' in 'C:\nwjs471'

Apply devtools.patch in C:\nwjs471\src\third_party\devtools-frontend\src
        1       0       front_end/extensions/ExtensionServer.js
... successfully applied.

Apply selenium.patch in C:\nwjs471\src\third_party\webdriver\pylib
        21      0       selenium/webdriver/chrome/options.py
... failed to apply:
        error: patch failed: selenium/webdriver/chrome/options.py:25
        error: selenium/webdriver/chrome/options.py: patch does not apply
-------------------------------------------------------------------------------
!!!! ERROR: This patch failed to apply. Your build will not be correct.
-------------------------------------------------------------------------------

Apply ffmpeg.patch in C:\nwjs471\src\third_party\ffmpeg
        8       4       BUILD.gn
... failed to apply:
        warning: BUILD.gn has type 100644, expected 100755
        error: patch failed: BUILD.gn:264
        error: BUILD.gn: patch does not apply
-------------------------------------------------------------------------------
!!!! ERROR: This patch failed to apply. Your build will not be correct.
-------------------------------------------------------------------------------

Apply vulcan.patch in C:\nwjs471\src\third_party\angle\third_party\vulkan-loader\src
        2       2       BUILD.gn
... successfully applied.

Apply icu.patch in C:\nwjs471\src\third_party\icu
        10      0       icu.gyp
... failed to apply:
        error: patch failed: icu.gyp:43
        error: icu.gyp: patch does not apply
-------------------------------------------------------------------------------
!!!! ERROR: This patch failed to apply. Your build will not be correct.
-------------------------------------------------------------------------------

Apply perfetto.patch in C:\nwjs471\src\third_party\perfetto
        1       0       src/trace_processor/importers/proto/packet_sequence_state.cc
... failed to apply:
        error: patch failed: src/trace_processor/importers/proto/packet_sequence_state.cc:14
        error: src/trace_processor/importers/proto/packet_sequence_state.cc: patch does not apply
-------------------------------------------------------------------------------
!!!! ERROR: This patch failed to apply. Your build will not be correct.
-------------------------------------------------------------------------------

6 patches total (2 applied, 0 skipped, 4 failed)

-------------------------------------------------------------------------------
!!!! ERROR: 4 patches failed to apply. Your build will not be correct.
-------------------------------------------------------------------------------
Traceback (most recent call last):
  File "src/content/nw/tools/patcher.py", line 121, in <module>
    apply_patch_config()
  File "src/content/nw/tools/patcher.py", line 96, in apply_patch_config
    results['fail'])
Exception: 4 patches failed to apply. Your build will not be correct.
use_relative_hooks is deprecated, please remove it from DEPS. (it was merged in use_relative_paths)
use_relative_hooks is deprecated, please remove it from DEPS. (it was merged in use_relative_paths)
Error: Command 'vpython.bat src/content/nw/tools/patcher.py' returned non-zero exit status 1 in C:\nwjs471

### How to reproduce

Went through the guide of :
http://docs.nwjs.io/en/latest/For%20Developers/Building%20NW.js/

and used nw47 branch, for v8, node, and nw.jw

when ran the gclient sync --with_branch_heads i got the error above.

the command "where python" output:

C:\Users\Person>where python
C:\depot_tools\python.bat
C:\Python27\python.exe
C:\Python38\python.exe
C:\Users\Person\AppData\Local\Microsoft\WindowsApps\python.exe

**Comment by rogerwang** [comment]
I don't think gclient supports MINGW64 environment.

**Comment by KanomiDim** [comment]
I'm using regular git bash. What should i use instead?

**Comment by maierfelix** [comment]
I have the same problem

**Comment by rogerwang** [comment]
> I'm using regular git bash. What should i use instead?

Just windows command line.

**Comment by rogerwang** [comment]
close until more information is available.

**Comment by fantasydr** [comment]
I met the same issue but managed to find a workaround.
TLDR: patch failed because of the EOL mismatch bwteen patch and target file (\r\n VS \n) on windows.

There is a special logic in nwjs\src\content\nw\tools\git_util.py to handle this issue:
```
def git_apply_patch_file(patch_path, patch_dir):
...
  if sys.platform == 'win32':
    # Convert the patch to Unix line endings. This is necessary to avoid
    # whitespace errors with git apply.
    patch_string = patch_string.replace('\r\n', '\n')
```
Basically it assume the patch file might contain '\r\n' but target file should always be '\n'.
This assumption seems wrong because one of the target file in vulcan.patch: 
```nwjs\src\third_party\angle\third_party\vulkan-loader\src\BUILD.gn```
It uses '\r\n' for EOL symbol on my machine after synced.

For the workaround, user could comment out the EOL replacement code in 'git_util.py'. Then manually modify all EOL in patch files in 'nwjs\src\content\nw\patch\patches' folder except 'vulcan.patch', this one should keep '\r\n'.

After this workaround, in cmd, run this in nwjs\ folder to resync everything:
```vpython.bat src/content/nw/tools/sync.py```

A possible solution could be check the target file and patch to make sure the EOL is matched.

**Comment by KanomiDim** [comment]
Hello everyone.

I solved this issues a few years ago (2020), but i didn't document on how i did that.

Unfortunately, a new computer and the problem came back..

So after digging alot, i was able to solve this after re-installing the Git for windows, but when the installer gui is running in the page of CRLF or how to checkout the code or something like this, i chose the third option to checkout and checkin as is (auto crlf set to false)


In addition, i fixed and installed VS Community edition 2017 + 2019 + 2022, including in the additional components of Windows SDK, and in Windows SDK I added the component of Debbuging tools for windows (you can access this through "Programs and Features" and then right click Windows SDK and choose Change, then choose Change and then you can add this component)


Hope I could help some desperate people that got all the way to here because of this issue.
